mm2src/mm2_main/src/rpc/lp_commands/lr_swap/lr_impl.rs

```diff
@@ -0,0 +1,350 @@
+        for (key, val) in self.inner.iter_mut() {
+            val.lr_swap_data = lr_swap_data.remove(key);
+        }
+    }
+
+    async fn update_with_contracts(&mut self, ctx: &MmArc) -> MmResult<(), ApiIntegrationRpcError> {
+        for ((src_token, dst_token), lr_data) in self.inner.iter_mut() {
+            let (src_coin, src_contract) = get_coin_for_one_inch(ctx, src_token).await?;
+            let (dst_coin, dst_contract) = get_coin_for_one_inch(ctx, dst_token).await?;
+            let src_decimals = src_coin.decimals();
+            let dst_decimals = dst_coin.decimals();
+            if src_decimals == 0 {
+                return MmError::err(ApiIntegrationRpcError::InternalError(format!(
+                    "Source token '{}' has invalid decimals (0)",
+                    src_token
+                )));
+            }
+            if dst_decimals == 0 {
+                return MmError::err(ApiIntegrationRpcError::InternalError(format!(
+                    "Destination token '{}' has invalid decimals (0)",
+                    dst_token
+                )));
+            }
+            lr_data.src_contract = Some(src_contract);
+            lr_data.dst_contract = Some(dst_contract);
+            lr_data.src_decimals = Some(src_decimals);
+            lr_data.dst_decimals = Some(dst_decimals);
+            lr_data.chain_id = Some(dst_coin.chain_id());
+        }
+        Ok(())
+    }
+
+    /// Query 1inch token_0/token_1 price in series and calc average price
+    /// Assuming the outer RPC-level code ensures that relation src_tokens : dst_tokens will never be M:N (but only 1:M or M:1)
+    async fn query_destination_token_prices(&mut self, ctx: &MmArc) -> MmResult<(), ApiIntegrationRpcError> {
+        let mut prices_futs = vec![];
+        let mut src_dst = vec![];
+        for ((src_token, dst_token), lr_data) in self.inner.iter() {
+            let (src_contract, dst_contract, chain_id) = lr_data.get_chain_contract_info()?;
+            // Run src / dst token price query:
+            let query_params = CrossPriceParams::new(chain_id, src_contract, dst_contract)
+                .with_granularity(Some(CROSS_PRICES_GRANULARITY))
+                .with_limit(Some(CROSS_PRICES_LIMIT))
+                .build_query_params()?;
+            let url_builder = PortfolioUrlBuilder::create_api_url_builder(ctx, PortfolioApiMethods::CrossPrices)?;
+            let fut = ApiClient::call_one_inch_api::<CrossPricesSeries>(url_builder, Some(query_params));
+            prices_futs.push(fut);
+            src_dst.push((src_token.clone(), dst_token.clone()));
+        }
+        let prices_in_series = join_all(prices_futs)
+            .await
+            .into_iter()
+            .filter_map(|res| if let Ok(prices) = res { Some(prices) } else { None }) // skip if bad result received (for e.g. low liguidity)
+            .collect::<Vec<_>>();
```

Comment:
nit: there is no need to collect this into `Vec<_>` and then `.iter()` it again in the next usage. i think.